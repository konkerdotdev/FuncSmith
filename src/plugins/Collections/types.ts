import type { FileSetItem } from '../../lib/fileSet';

export type Collection<T extends FileSetItem> = T & {
  readonly collection: {
    readonly len: number;
    readonly previous?: number;
    readonly next?: number;
  };
};

export type CollectionOptions = {
  readonly globPattern: string;
  readonly sortBy: string;
  readonly reverse: boolean;
};

export type Convenience<T extends Partial<CollectionOptions>> = string | T;
