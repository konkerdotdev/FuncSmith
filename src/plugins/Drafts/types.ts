export type DraftsOptions = {
  readonly default: boolean;
  readonly include: boolean;
};
