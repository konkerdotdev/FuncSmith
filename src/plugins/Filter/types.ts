export type FilterOptions = {
  // An array of glob patterns, of files to drop if matched
  readonly drop: Array<string>;

  // An array of glob patterns, of files to keep if matched
  readonly keep: Array<string>;
};
