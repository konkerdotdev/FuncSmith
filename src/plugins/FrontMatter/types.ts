import type { FileSetItem } from '../../lib/fileSet';

export type FrontMatter<T extends Partial<FileSetItem>> = T & {
  readonly frontMatter: Record<string, unknown>;
};

export type FrontMatterOptions = {
  readonly globPattern: string;
};
