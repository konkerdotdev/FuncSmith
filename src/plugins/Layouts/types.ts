import type * as H from 'handlebars';

export type LayoutsOptions = {
  readonly templateEngine: string;
  readonly directory: string;
  readonly defaultLayout: string;
  readonly globPattern: string;
  readonly helpers: Record<string, H.HelperDelegate>;
};
