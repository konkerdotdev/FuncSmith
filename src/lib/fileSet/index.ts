import * as P from '@konker.dev/effect-ts-prelude';

import type { FrontMatter } from '../frontMatter';

export enum FileSetItemType {
  File = 'File',
}

export type FileSetItemFile<T = ArrayBuffer | string> = {
  _tag: FileSetItemType.File;
  path: string;
  baseDir: string;
  relPath: string;
  relDir: string;
  fileName: string;
  fileBase: string;
  fileExt: string;
  contents: T;
};

export type FileSetItem<T = ArrayBuffer | string> = FileSetItemFile<T>;

export type FileSet<T extends FileSetItem<C>, C = ArrayBuffer | string> = Array<T>;

// --------------------------------------------------------------------------
export const HtmlString = P.pipe(P.Schema.string, P.Schema.brand(Symbol.for('HtmlString')));
export type HtmlString = P.Schema.Schema.To<typeof HtmlString>;
export type Html<T extends FileSetItem> = FrontMatter<Omit<T, 'contents'>> & {
  readonly contents: HtmlString;
};
