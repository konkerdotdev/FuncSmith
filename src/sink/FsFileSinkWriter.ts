import * as P from '@konker.dev/effect-ts-prelude';

import type { FileSet } from '../mappings';
import type { FileSink, FileSinkWriter } from './index';

export const fsFileSinkWriter: FileSinkWriter = (_fs: FileSink, _d: FileSet) => {
  console.log('KONK90', _d);
  // Write the list of FileItems to the file system at the given path
  return P.Effect.succeed(P.Effect.unit);
};
