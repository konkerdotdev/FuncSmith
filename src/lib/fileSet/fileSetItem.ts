import * as P from '@konker.dev/effect-ts-prelude';
import type { TinyFileSystem, TinyFileSystemError } from '@konker.dev/tiny-filesystem-fp';
import { arrayBufferToString } from '@konker.dev/tiny-filesystem-fp/dist/lib/array';
import type { DirectoryData, FileData } from '@konker.dev/tiny-treecrawler-fp';
import { isFileData } from '@konker.dev/tiny-treecrawler-fp/dist/lib/utils';
import micromatch from 'micromatch';

import type { FuncSmithError } from '../../error';
import { toFuncSmithError } from '../../error';
import type { GeneralError } from '../utils';
import { hashHex } from '../utils';
import type { FileSetItem, FileSetItemFile } from './index';
import { FileSetItemType } from './index';

export type RenameSpec = [RegExp, string];

// --------------------------------------------------------------------------
export function isFileSetItemFile<T extends FileSetItem>(data: T): data is T & FileSetItemFile {
  return data._tag === FileSetItemType.File;
}

// --------------------------------------------------------------------------
export function fileSetItemRename<T extends FileSetItemFile>(
  tfs: TinyFileSystem,
  renameSpec: RenameSpec,
  item: T
): P.Effect.Effect<never, FuncSmithError, T> {
  const relPath = item.relPath.replace(renameSpec[0], renameSpec[1]);
  const fileName = tfs.basename(relPath);
  const fileExt = tfs.extname(fileName);
  const fileBase = tfs.basename(fileName, fileExt);
  return P.pipe(
    P.Effect.Do,
    P.Effect.bind('path', () => tfs.joinPath(item.baseDir, relPath)),
    P.Effect.bind('relDir', () => tfs.dirName(relPath)),
    P.Effect.map(({ path, relDir }) => ({
      ...item,
      path,
      relPath,
      relDir,
      fileName,
      fileBase,
      fileExt,
    })),
    P.Effect.mapError(toFuncSmithError)
  );
}

// --------------------------------------------------------------------------
export function fileSetItemMatchesPattern<T extends FileSetItemFile>(
  globPattern: string | undefined,
  fileSetItem: T
): boolean {
  return !!globPattern ? micromatch([fileSetItem.path], [globPattern])?.length > 0 : true;
}

// --------------------------------------------------------------------------
export const toFileSetItemFile =
  (tfs: TinyFileSystem, sourcePath: string) =>
  (i: FileData): P.Effect.Effect<never, TinyFileSystemError | GeneralError, FileSetItemFile> => {
    const fileName = tfs.basename(i.path);
    const fileExt = tfs.extname(i.path);

    return P.pipe(
      P.Effect.Do,
      P.Effect.bind('relDir', () => tfs.dirName(tfs.relative(sourcePath, i.path))),
      P.Effect.bind('pathHash', () => hashHex(i.path)),
      P.Effect.map(({ pathHash, relDir }) => ({
        _tag: FileSetItemType.File,
        _id: pathHash,
        path: i.path,
        baseDir: sourcePath,
        relPath: tfs.relative(sourcePath, i.path),
        relDir,
        fileName,
        fileExt,
        fileBase: tfs.basename(i.path, fileExt),
        contents: new Uint8Array(i.data),
      }))
    );
  };

// --------------------------------------------------------------------------
export const toFileSystemItemList =
  (tfs: TinyFileSystem, sourcePath: string) =>
  (
    list: Array<DirectoryData | FileData>
  ): P.Effect.Effect<never, TinyFileSystemError | GeneralError, Array<FileSetItem>> =>
    P.pipe(list.filter(isFileData), P.Array.map(toFileSetItemFile(tfs, sourcePath)), P.Effect.all);

// --------------------------------------------------------------------------
export function fileSetItemContentsToString(contents: ArrayBuffer | string): string {
  return typeof contents === 'string' ? contents : arrayBufferToString(contents);
}
