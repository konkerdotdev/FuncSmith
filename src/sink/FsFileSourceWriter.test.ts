import * as P from '@konker.dev/effect-ts-prelude';
import { MemFsTinyFileSystem } from '@konker.dev/tiny-filesystem-fp/dist/memfs';

import * as fileSetItemLib from '../lib/fileSet/fileSetItem';
import * as fixturesFs from '../test/fixtures/fileset-1';
import fixturesMemFs1 from '../test/fixtures/memfs-1.json';
import { FsDepTinyFileSystem } from '../types';
import * as unit from './FsFileSinkWriter';

const TEST_TFS = MemFsTinyFileSystem(fixturesMemFs1, '/tmp');
import { fs } from 'memfs';
import { toTreeSync } from 'memfs/lib/print';

describe('sink', () => {
  describe('FsFileSinkWriter', () => {
    beforeEach(() => {
      jest.spyOn(fileSetItemLib, 'isFileSetItemFile').mockReturnValueOnce(false);
    });
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should write files to the file system', async () => {
      const sinkPath = '/tmp/foo';
      await P.Effect.runPromise(
        P.pipe(
          unit.fsFileSinkWriter(sinkPath, fixturesFs.TEST_FILE_SET_1),
          P.Effect.provideService(
            FsDepTinyFileSystem,
            FsDepTinyFileSystem.of({
              tinyFs: TEST_TFS,
            })
          )
        )
      );

      // Inspect state of memfs
      expect(toTreeSync(fs)).toMatchSnapshot('fsFileSinkWriter_memfs_tree_1');
    });
  });
});
