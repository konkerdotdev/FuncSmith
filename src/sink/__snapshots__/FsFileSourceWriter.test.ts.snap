// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sink FsFileSinkWriter should write files to the file system: fsFileSinkWriter_memfs_tree_1 1`] = `
"/
└─ tmp/
   ├─ foo/
   │  ├─ a.txt
   │  ├─ b.txt
   │  ├─ bar/
   │  │  ├─ e.txt
   │  │  └─ f.log
   │  ├─ c.csv
   │  ├─ d.json
   │  ├─ ignore1.txt
   │  └─ posts/
   │     ├─ index.md
   │     ├─ p1.md
   │     └─ p2.md
   ├─ layouts/
   │  ├─ bar.hbs
   │  ├─ foo.hbs
   │  ├─ layout.hbs
   │  └─ post.hbs
   └─ partials/
      └─ footer.hbs"
`;
